### DISCUSSION TITLE
Issue running `tar_make_clustermq()` via ssh?

### DISCUSSION THREAD

**Discussion by JasonPekos** [ropensci/targets#1078] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Hi! I've been running into an issue trying to run targets with `tar_make_clustermq()`. As an example, when running the simple example pipeline, I get the following error:

```r
library(targets)

tar_option_set(
  packages = c("tibble"),
  format = "rds"
)

options(clustermq.scheduler = "ssh")
options(clustermq.ssh.host = "user@host")

list(
  tar_target(
    name = data,
    command = tibble(x = rnorm(100), y = rnorm(100))
  )
)
```

```
Error:
! Error running targets::tar_make_clustermq()
  Target errors: targets::tar_meta(fields = error, complete_only = TRUE)
  Tips: https://books.ropensci.org/targets/debugging.html
  Last error: $ operator is invalid for atomic vectors
 ```

---

## Debugging: Is ClusterMQ working?

Yes, it looks like ClusterMQ is working. I can use the example ClusterMQ code:

```r
library(clustermq)

fx = function(x) x * 2
Q(fx, x=1:3, n_jobs=1)
```

and ClusterMQ will send the jobs to relevant cluster and return the result. 

## Debugging: Does `tar_make_clustermq()` work locally on the cluster, e.g. without SSH? 

Yes, everything works fine on the cluster with `slurm`, and on my computer with `multicore` as the scheduler. The issue only seems to pop up with `ssh` as the scheduler.

## What does my setup look like?

**On my local computer** I have the following `.Rprofile`:

```r
options(
    clustermq.scheduler = "ssh",
    clustermq.ssh.host = "user@host", 
    clustermq.ssh.log = "~/cmq_ssh.log" 
)
```

**On my cluster** I have the following `.Rprofile`:

```r
options(
    clustermq.template = "~/cmq.tmpl" # template for the scheduler script
)
```

Where the relevant template is:

```sh
#!/bin/sh

# Modified from https://github.com/mschubert/clustermq/blob/master/inst/SLURM.tmpl
# under the Apache 2.0 license.

#SBATCH --job-name=superconservative_halcyon_{{ job_name }}
#SBATCH --output={{ log_file | /dev/null }}
#SBATCH --error={{ log_file | /dev/null }}
#SBATCH --mem-per-cpu={{ memory | 5G }}
#SBATCH --array=1-{{ n_jobs }}
#SBATCH --cpus-per-task={{ cores | 4 }}
#SBATCH --account=def-user 
#SBATCH --time=0-01:00

module load R # Comment out if R is not an environment module.
# ulimit -v $(( 1024 * {{ memory | 3G }} ))
CMQ_AUTH={{ auth }} R --no-save --no-restore -e 'clustermq:::worker("{{ master }}")'
```

I also load the R module I want in my `.bashrc`:

```
module load r/4.2.2
```

**On my local computer, in my _targets.R file** I have the following options (same as mentioned at the start of this discussion):

```r
library(targets)

# Set target options:
tar_option_set(
  packages = c("tibble"),
  format = "rds"
)

options(clustermq.scheduler = "ssh")
options(clustermq.ssh.host = "pekos@graham.computecanada.ca")
```

## More about the issue:

This is everything I see when I run `tar_make_clustermq`

```
> tar_make_clustermq(workers = 2)
Connecting pekos@graham.computecanada.ca via SSH ...
Sending common data ...
• start target data
• end pipeline [1.278 minutes]
Error:
! Error running targets::tar_make_clustermq()
  Target errors: targets::tar_meta(fields = error, complete_only = TRUE)
  Tips: https://books.ropensci.org/targets/debugging.html
  Last error: $ operator is invalid for atomic vectors
> 
```

So, I can connect to the cluster, however something seems to go wrong when running the pipeline.

**Comment by wlandau** [comment, marked as answer]
> Yes, it looks like ClusterMQ is working. I can use the example ClusterMQ code:
> 
> library(clustermq)
> 
> fx = function(x) x * 2
> Q(fx, x=1:3, n_jobs=1)

Assuming your `clustermq`-only example implicitly includes `options(clustermq.scheduler = "ssh")` , etc., which I assume it does, it's strange that this works but the `targets` example does not.

I wonder if it is as simple as different versions of `targets` vs the cluster. Would you check that both the cluster and your local machine have the same `targets` version?

Unfortunately my company blocks SSH tunneling, so I cannot use the SSH connector at all. But if a package version mismatch does not explain it, you could try to get a traceback and see where the error happens:

```r
remotes::install_github("ropensci/targets", ref = "1078")
targets::tar_make(callr_function = NULL)
traceback()
```

**Reply by JasonPekos** [reply]
I know this is an annoying / unsatisfactory answer, but I ended up reinstalling all the relevant packages (targets / clustermq) both locally and in my module on the cluster. It looks like everything is working now. 

I think you're right about it being some sort of version error — I did update my packages earlier during debugging when I was warned about a clustermq version conflict, but very possible that something went wrong.

**Reply by wlandau** [reply]
Glad you figured it out.
